File: gathering-api/src/main/java/gg/warcraft/gathering/api/gatherable/BlockGatherable.java
package gg.warcraft.gathering.api.gatherable;

import gg.warcraft.monolith.api.world.block.Block;
import gg.warcraft.monolith.api.world.block.BlockTypeVariantOrState;

/**
 * A BlockGatherable represents a block that can be harvested for resources such as a STONE block in a mine or a WOOD
 * block in a forest. Gatherables respawn after a cooldown.
 */
public interface BlockGatherable extends Gatherable {

    /**
     * @return The type this gatherable appears as while on cooldown. Never null.
     */
    BlockTypeVariantOrState getCooldownBlockData();

    /**
     * @param block The type. Can not be null.
     * @return True if this gatherable contains the type, false otherwise.
     */
    boolean containsBlock(Block block);
}

File: gathering-api/src/main/scala/gg/warcraft/gathering/api/gatherable/BlockGatherable.scala
package gg.warcraft.gathering.api.gatherable

import gg.warcraft.monolith.api.core.Duration
import gg.warcraft.monolith.api.world.WorldService
import gg.warcraft.monolith.api.world.block.{Block, BlockTypeVariantOrState}
import gg.warcraft.monolith.api.world.item.{Item, ItemService}

import scala.util.Random

class BlockGatherable(
    blockDataString: String,
    dropDataString: String,
    dropName: String,
    cooldown: Int,
    cooldownDelta: Int,
    cooldownDataString: String
)(
    private implicit val itemService: ItemService,
    private implicit val worldService: WorldService
) {
  private val blockData = worldService.parseData(blockDataString)
  private val dropData = itemService.parseData(dropDataString)

  val cooldownData: BlockTypeVariantOrState =
    worldService.parseData(cooldownDataString)

  def matches(block: Block): Boolean =
    block.hasData(blockData)

  def generateDrop: Item =
    itemService.create(dropData).withName(dropName)

  def generateCooldown: Duration =
    Duration.ofSeconds(cooldown + Random.nextInt(cooldownDelta))
}
